beluga_tutorial/src/main.cpp

```diff
@@ -0,0 +1,258 @@
+      node["simulation_dataset"][cycle]["particles"]["current"]["weights"].push_back(
+          static_cast<double>(std::get<1>(current)));
+    }
+
+    for (const auto& propagate : data.propagate) {
+      node["simulation_dataset"][cycle]["particles"]["propagate"]["states"].push_back(std::get<0>(propagate));
+      node["simulation_dataset"][cycle]["particles"]["propagate"]["weights"].push_back(
+          static_cast<double>(std::get<1>(propagate)));
+    }
+
+    for (const auto& reweight : data.reweight) {
+      node["simulation_dataset"][cycle]["particles"]["reweight"]["states"].push_back(std::get<0>(reweight));
+      node["simulation_dataset"][cycle]["particles"]["reweight"]["weights"].push_back(
+          static_cast<double>(std::get<1>(reweight)));
+    }
+
+    for (const auto& resample : data.resample) {
+      node["simulation_dataset"][cycle]["particles"]["resample"]["states"].push_back(std::get<0>(resample));
+      node["simulation_dataset"][cycle]["particles"]["resample"]["weights"].push_back(
+          static_cast<double>(std::get<1>(resample)));
+    }
+
+    node["simulation_dataset"][cycle]["estimation"]["mean"] = std::get<0>(data.estimation);
+    node["simulation_dataset"][cycle]["estimation"]["sd"] = std::get<1>(data.estimation);
+  }
+
+  const std::string bags_path{"./src/beluga/beluga_tutorial/bags/"};
+  std::filesystem::create_directory(bags_path);
+  std::ofstream fout(bags_path + file_name, std::ios::trunc);
+  fout << node;
+  fout.close();
+}
+
+int main() {
+  // Load params from yaml file
+  TutorialParams tutorial_params;
+  load_params_from_yaml(tutorial_params);
+
+  // Create the TutorialDataset object for record the simulation data
+  TutorialDataset tutorial_dataset;
+
+  // Load map from file or generate a random map
+  LandmarkMapVector landmark_map;
+  if (tutorial_params.load_landmark_map_from_file) {
+    load_landmark_map_from_yaml(landmark_map);
+  } else {
+    std::uniform_int_distribution landmark_distribution{0, tutorial_params.map_size};
+    landmark_map = beluga::views::sample(landmark_distribution) |
+                   ranges::views::take_exactly(tutorial_params.number_of_doors) | ranges::to<LandmarkMapVector>;
+    landmark_map |= ranges::actions::sort | ranges::actions::unique;
+  }
+
+  tutorial_dataset.landmark_map = landmark_map;
+
+  // Generate particles
+  std::normal_distribution<double> initial_distribution(tutorial_params.initial_pose, tutorial_params.initial_pose_sd);
+  auto particles = beluga::views::sample(initial_distribution) |
+                   ranges::views::transform(beluga::make_from_state<Particle>) |
+                   ranges::views::take_exactly(tutorial_params.number_of_particles) | ranges::to<beluga::TupleVector>;
+
+  // Execute the particle filter using beluga
+  double current_pose{tutorial_params.initial_pose};
+  for (auto n = 0; n < tutorial_params.number_of_cycles; n++) {
+    // Prepare the data to be saved
+    TutorialData tutorial_data;
```

Comment:
@Alondruck meta: could we reuse the same `TutorialData` instance on each iteration? It look like we overwrite it in full.

Reply:
I thought it would be more efficient, or actually more correct this way, since at the moment I leave the simulation loop, tutorial_data has no use.However, I can reuse it by defining it only once outside.